### PR TITLE
Adding ability to load model from string in R

### PR DESCRIPTION
**Pull request by JesseLimtiaco**
Changes to lgb.Booster.R:
Added method to call LGBM_BoosterLoadModelFromString_R if model_str is provided for initialization, added option to load from model_str in lgb.load

Changes to lightgbm_r.cpp:  Added LGBM_BoosterLoadModelFromString_R to call the C api

### CODE REVIEW

**Comment by guolinke** [comment]
@JesseLimtiaco Thanks!
BTW, it seems you also need to have ```LGBM_BoosterSaveModelToString_R```.

**Comment by JesseLimtiaco** [comment]
OK, I implemented LGBM_BoosterSaveModelToString_R and the save_model_to_string method in the 2 latest commits

**Comment by guolinke** [comment]
@JesseLimtiaco can you sign the CLA.

@Laurae2 can this also work for the saveRDS? I think we not need to save model to file when save R object any more.

**Comment by guolinke** [comment]
@JesseLimtiaco  did this solve #432 ?

File: R-package/R/lgb.Booster.R [hunk 1]
```diff
@@ -343,6 +356,21 @@ Booster <- R6Class(
...
+    # Save model to string
+    save_model_to_string = function(num_iteration = NULL) {
+      
+      # Check if number of iteration is non existent
+      if (is.null(num_iteration)) {
+        num_iteration <- self$best_iter
+      }
+      
+      # Return model string
+      lgb.call.return.str("LGBM_BoosterSaveModelToString_R",
```
Comment (Laurae2) [inline R-package/R/lgb.Booster.R]: Add `return` here for readability

File: R-package/R/lgb.Booster.R [hunk 2]
```diff
@@ -671,19 +699,29 @@ predict.lgb.Booster <- function(object, data,
 #'                    early_stopping_rounds = 10)
 #' lgb.save(model, "model.txt")
 #' load_booster <- lgb.load("model.txt")
+#' load_booster <- lgb.load(model_str)
 #' }
 #' 
 #' @rdname lgb.load
 #' @export
 lgb.load <- function(filename){
```
Comment (Laurae2) [inline R-package/R/lgb.Booster.R]: You should add a way to check difference between `filename` and `model_str`.

File: R-package/R/lgb.Booster.R [hunk 3]
```diff
@@ -671,19 +699,29 @@ predict.lgb.Booster <- function(object, data,
...
   
   # Check if file name is character or not
-  if (!is.character(filename)) {
+  if (supplied(filename) && !is.null(filename) && !is.character(filename)) {
     stop("lgb.load: filename should be character")
   }
-  
+    
+  if (supplied(model_str) && !is.null(model_str) && !is.character(model_str)) {
```
Comment (Laurae2) [inline R-package/R/lgb.Booster.R]: model_str undefined, remove it.

File: R-package/R/lgb.Booster.R [hunk 4]
```diff
@@ -671,19 +699,29 @@ predict.lgb.Booster <- function(object, data,
...
+    stop("lgb.load: model_str should be character")
+  }
+    
+  if (!supplied(filename) && !supplied(model_str)) {
```
Comment (Laurae2) [inline R-package/R/lgb.Booster.R]: model_str undefined, remove it.

File: R-package/R/lgb.Booster.R [hunk 5]
```diff
@@ -671,19 +699,29 @@ predict.lgb.Booster <- function(object, data,
...
+    stop("lgb.load: either filename or model_str must be supplied")
+  }
+    
   # Return new booster
-  Booster$new(modelfile = filename)
+  if (supplied(filename)) Booster$new(modelfile = filename)
+  if (supplied(model_str)) Booster$new(model_str = model_str)
```
Comment (Laurae2) [inline R-package/R/lgb.Booster.R]: model_str undefined, remove it.

File: R-package/R/lgb.Booster.R [hunk 6]
```diff
@@ -671,19 +699,29 @@ predict.lgb.Booster <- function(object, data,
...
+    stop("lgb.load: model_str should be character")
+  }
+    
+  if (!supplied(filename) && !supplied(model_str)) {
+    stop("lgb.load: either filename or model_str must be supplied")
+  }
+    
   # Return new booster
-  Booster$new(modelfile = filename)
+  if (supplied(filename)) Booster$new(modelfile = filename)
```
Comment (Laurae2) [inline R-package/R/lgb.Booster.R]: You can use `file.exists` to check for `filename` file existence.

If file exists, you can load as usual from file. Otherwise, you can attempt to load model from string. You may separate loading from string and loading from model if needed, it requires a different argument apparently in your implementation (model_str).

**Comment by Laurae2** [comment]
@guolinke Yes it can work, but first we have to modify `readRDS.lgb.booster` and `saveRDS.lgb.Booster`. I can implement that after the PR author fixes the default package checks and consistency (and it gets merged afterwards).

@JesseLimtiaco Also, can you provide a working example?

File: R-package/R/lgb.Booster.R [hunk 7]
```diff
@@ -671,19 +699,29 @@ predict.lgb.Booster <- function(object, data,
 #'                    early_stopping_rounds = 10)
 #' lgb.save(model, "model.txt")
 #' load_booster <- lgb.load("model.txt")
+#' load_booster <- lgb.load(model_str)
```
Comment (Laurae2) [inline R-package/R/lgb.Booster.R]: model_str does not exist in the example.

**Comment by JesseLimtiaco** [comment]
Thanks for all the comments @Laurae2 , I'm still getting used to working in github.  I have updated the files and below is an example of the ways the new loading/saving should work in R:

require(lightgbm)
require(methods)
data(agaricus.train, package = "lightgbm")
data(agaricus.test, package = "lightgbm")
train <- agaricus.train
test <- agaricus.test
class(train$label)
class(train$data)


bst <- lightgbm(data = train$data,
                label = train$label,
                num_leaves = 4,
                learning_rate = 1,
                nrounds = 2,
                objective = "binary")

filename <- tempfile()
bst$save_model(filename = filename)
model_load_1 <- lgb.load(filename = filename)

bst$save()
model_str <- bst$save_model_to_string()

model_load_2 <- lgb.load(model_str = bst$raw)
model_load_3 <- lgb.load(model_str = model_str)

**Comment by JesseLimtiaco** [comment]
I have tested the above example and @guolinke  this resolves #432

**Comment by guolinke** [comment]
OK, i will merge this. 
@Laurae2 you can fix it later.
